common/net/credentials/alts/handshaker/handshaker.go

```diff
@@ -0,0 +1,408 @@
+		PastelID:  signInfo.PastelID,
+	}
+	// encode and write the client's handshake record
+	if err := s.writeHandshake(&request); err != nil {
+		return fmt.Errorf("write handshake: %w", err)
+	}
+
+	// read and decode the client's handshake record
+	var response kexExchangeResponse
+	if err := s.readHandshake(&response); err != nil {
+		return fmt.Errorf("read handshake: %w", err)
+	}
+
+	if ok, err := secClient.Verify(ctx, append(response.PubKey, challengeA...),
+		string(response.Signature), response.PastelID); err != nil || !ok {
+		return fmt.Errorf("failed to verify server public key: %w", err)
+	}
+
+	var serverPubKey [ED448PubKeySize]byte
+	copy(serverPubKey[:], response.PubKey)
+	// compute the secret key for connection
+	secret := curve.ComputeSecret(priv, serverPubKey)
+	if len(secret) < cryptokdf.CryptoKdfKeybytes() {
+		return fmt.Errorf("secret key length is missmatch, len=%d", len(secret))
+	}
+
+	subkeyLen, ok := recordKeyLen[s.protocol]
+	if !ok {
+		return fmt.Errorf("unknown resulted record protocol: %v", s.protocol)
+	}
+
+	kdfValue, exit := cryptokdf.CryptoKdfDeriveFromKey(subkeyLen, kdfSubKeyID, kdfContext,
+		secret[:cryptokdf.CryptoKdfKeybytes()])
+	if exit != 0 {
+		return fmt.Errorf("failed to compute kdf, exit=%d", exit)
+	}
+	// update the record key
+	s.key = kdfValue
+
+	return nil
+}
+
+// ClientHandshake starts and completes a client ALTS handshaking. Once
+// done, ClientHandshake returns a secure connection.
+func (s *altsHandshaker) ClientHandshake(ctx context.Context, secClient alts.SecClient, signInfo *alts.SignInfo) (net.Conn, credentials.AuthInfo, error) {
+	if !acquire() {
+		return nil, nil, errDropped
+	}
+	defer release()
+
+	if s.side != alts.ClientSide {
+		return nil, nil, errors.New("only handshakers created using NewClientHandshaker can perform a client handshaker")
+	}
+
+	// do the client handshake
+	if err := s.doClientHandshake(ctx, secClient, signInfo); err != nil {
+		return nil, nil, fmt.Errorf("do client handshake: %w", err)
+	}
+	log.WithContext(ctx).Debugf("client handshake is complete")
+
+	// new a secure connection
+	sc, err := conn.NewConn(s.side, s.conn, s.protocol, s.key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new secure conn: %w", err)
+	}
+
+	return sc, authinfo.New(), nil
+}
+
+func (s *altsHandshaker) doServerHandshake(ctx context.Context, secClient alts.SecClient, signInfo *alts.SignInfo) error {
+	var challengeA []byte
+	if err := s.readHandshake(&challengeA); err != nil {
+		return fmt.Errorf("read handshake: %w", err)
+	}
+
+	challengeB := make([]byte, challengeSize)
+	rand.Read(challengeB)
+	if err := s.writeHandshake(&challengeB); err != nil {
+		return fmt.Errorf("write handshake: %w", err)
+	}
+
+	// read and decode the client's handshake record
+	var request kexExchangeRequest
+	if err := s.readHandshake(&request); err != nil {
+		return fmt.Errorf("read handshake: %w", err)
+	}
+
+	if ok, err := secClient.Verify(ctx, append(request.PubKey, challengeB...),
```

Comment:
Preallocation is better here. `append(request.Pubkey, chanllengeB...)` must go through a lot of reallocation if the capacity of `request.Pubkey` is not enough.